Sources/TSFFutures/BatchingFutureOperationQueue.swift

```diff
@@ -51,49 +84,72 @@ public struct LLBBatchingFutureOperationQueue {
     ///    - group:     Threads capable of running futures.
     ///    - maxConcurrentOperationCount:
     ///                 Operations to execute in parallel.
-    @inlinable
-    public init(name: String, group: LLBFuturesDispatchGroup, maxConcurrentOperationCount maxOpCount: Int, qualityOfService: QualityOfService = .default) {
+    public init(name: String, group: LLBFuturesDispatchGroup, maxConcurrentOperationCount maxOpCount: Int) {
+        self.init(name: name, group: group, maxConcurrentOperationCount: maxOpCount, dispatchQoS: .default)
+    }
+
+    public init(
+        name: String, group: LLBFuturesDispatchGroup, maxConcurrentOperationCount maxOpCnt: Int,
+        dispatchQoS: DispatchQoS
+    ) {
         self.group = group
-        self.operationQueue = OperationQueue(tsf_withName: name, maxConcurrentOperationCount: maxOpCount)
-        self.operationQueue.qualityOfService = qualityOfService
+        self.name = name
+        self.qos = dispatchQoS
+
+        self.concurrencyLimiter = ConcurrencyLimiter(maximumConcurrency: Self.bridged(maxOpCount: maxOpCnt))
     }
 
-    @inlinable
     public func execute<T>(_ body: @escaping () throws -> T) -> LLBFuture<T> {
-        let promise = group.next().makePromise(of: T.self)
-        operationQueue.addOperation {
-            promise.fulfill(body)
+        return self.concurrencyLimiter.withReplenishableLimit(eventLoop: group.any()) { eventLoop in
+            let promise = eventLoop.makePromise(of: T.self)
+
+            DispatchQueue(label: self.name, qos: self.qos).async {
+                promise.fulfill(body)
+            }
+
+            return promise.futureResult
         }
-        return promise.futureResult
     }
 
-    @inlinable
     public func execute<T>(_ body: @escaping () -> LLBFuture<T>) -> LLBFuture<T> {
-        let promise = group.next().makePromise(of: T.self)
-        operationQueue.addOperation {
-            let f = body()
-            f.cascade(to: promise)
+        return self.concurrencyLimiter.withReplenishableLimit(eventLoop: group.any()) { eventLoop in
+            let promise = eventLoop.makePromise(of: T.self)
+
+            DispatchQueue(label: self.name, qos: self.qos).async {
```

Comment:
here you don't need to dispatch off. The `body` returns a future and therefore will be non-blocking.

Reply:
Yes, ideally that should be the case, but if I understand correctly, a rogue `body` could still do blocking operations? Executing in a dispatch queue would keep us more aligned with the previous implementation where we do not block the event loop in such a case. 

For e.g., 
```
batchFutOpQ.execute {
  someBlockingCall()
  return elg.makeSucceededFuture(())
}
```

(Note that this still blocks if there is a blocking call inside `flatMap` like `return elg.makeSucceededFuture(()).flatMap{ someBlockingCall() }`. But this would be the same behavior in previous implementation)

Reply:
Yes, that's true. But a rogue body could also just crash the program or worse. We shouldn't write overly defensive code and if we find such a rogue body, we should fix it instead. Every dispatch to another thread (and back) costs us performance so we should avoid it as much as possible.

Reply:
Understood now, makes sense. While I do agree we need to fix that, I think it is a non trivial task (pls correct me if I'm wrong here) to find and fix all usages. So I think it is better to do that as an enhancement as a different task since the objective of the task is to only get rid of using `OperationQueues`. Does that sound fair?

Reply:
Have you seen any code that uses this which is blocking and returns a future? There shouldn't really be any such code (of course there might but I'd hope it's not widespread).

Reply:
Discussed this on DM